src/Engines/Native/nativeDataStream.ts

```diff
@@ -0,0 +1,115 @@
+export interface NativeData extends Uint32Array {
+}
+
+interface INativeDataStream {
+    writeBytes(buffer: ArrayBuffer, byteLength: number): void;
+}
+
+declare var _native: any;
+
+export class NativeDataStream {
+    private readonly _data: DataView;
+    private readonly _uint32s: Uint32Array;
+    private readonly _int32s: Int32Array;
+    private readonly _float32s: Float32Array;
+    private _position: number;
+    private _nativeDataStream: INativeDataStream;
+
+    private static readonly DEFAULT_BUFFER_SIZE = 1024;
+
+    constructor () {
+        const buffer = new ArrayBuffer(NativeDataStream.DEFAULT_BUFFER_SIZE);
+        this._data = new DataView(buffer);
+        this._uint32s = new Uint32Array(buffer);
+        this._int32s = new Int32Array(buffer);
+        this._float32s = new Float32Array(buffer);
+
+        this._position = 0;
+
+        this._nativeDataStream = new _native.NativeDataStream(() => {
+            this._flush();
+        });
+    }
+
+    public writeUint8(value: number): void {
+        this._autoFlushIfNecessary(1);
+        this._data.setUint8(this._position, value);
+        this._position += 1;
+    }
+
+    public writeUint32(value: number): void {
+        this._autoFlushIfNecessary(4);
+        this._data.setUint32(this._position, value, true);
+        this._position += 4;
+    }
+
+    public writeInt32(value: number): void {
+        this._autoFlushIfNecessary(4);
+        this._data.setInt32(this._position, value, true);
+        this._position += 4;
+    }
+
+    public writeFloat32(value: number): void {
+        this._autoFlushIfNecessary(4);
+        this._data.setFloat32(this._position, value, true);
+        this._position += 4;
+    }
+
+    public writeUint32Array(values: Uint32Array): void {
+        this._byteAlign32();
+        this._autoFlushIfNecessary(4 + values.byteLength);

```

Comment:
adding 4 is for the alignment?

Reply:
No, we're already aligned by this call. The extra four bytes are for the array's length, which is written as a Uint32 before the array data.